File: gitdot-api/src/endpoint/migration/github/migrate_github_repositories.rs
use serde::{Deserialize, Serialize};

use crate::{endpoint::Endpoint, resource::migration::MigrationResource};

pub struct MigrateGitHubRepositories;

impl Endpoint for MigrateGitHubRepositories {
    const PATH: &'static str = "/migration/github/{installation_id}/migrate";
    const METHOD: http::Method = http::Method::POST;

    type Request = MigrateGitHubRepositoriesRequest;
    type Response = MigrateGitHubRepositoriesResponse;
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct MigrateGitHubRepositoriesRequest {
    pub origin: String,
    pub origin_type: String,
    pub destination: String,
    pub destination_type: String,
    pub repositories: Vec<String>,
    pub readonly: bool,
}

pub type MigrateGitHubRepositoriesResponse = MigrationResource;

File: gitdot-web/packages/gitdot-api/src/endpoint/migration/github/migrate-github-repositories.ts
import { z } from "zod";
import { MigrationResource } from "../../../resource";
import type { Endpoint } from "../../endpoint";

export const GitHubRepositoryRef = z.object({
  name: z.string(),
  id: z.number(),
});
export type GitHubRepositoryRef = z.infer<typeof GitHubRepositoryRef>;

export const MigrateGitHubRepositoriesRequest = z.object({
  origin: z.string(),
  origin_type: z.string(),
  destination: z.string(),
  destination_type: z.string(),
  repositories: z.array(GitHubRepositoryRef),
  readonly: z.boolean(),
});
export type MigrateGitHubRepositoriesRequest = z.infer<
  typeof MigrateGitHubRepositoriesRequest
>;

export const MigrateGitHubRepositoriesResponse = MigrationResource;
export type MigrateGitHubRepositoriesResponse = z.infer<
  typeof MigrateGitHubRepositoriesResponse
>;

export const MigrateGitHubRepositories = {
  path: "/migration/github/{installation_id}/migrate",
  method: "POST",
  request: MigrateGitHubRepositoriesRequest,
  response: MigrateGitHubRepositoriesResponse,
} as const satisfies Endpoint;
export type MigrateGitHubRepositories = typeof MigrateGitHubRepositories;
